califa/calibration/R3BCalifaCrystalCal2Cluster.cxx

```diff
@@ -138,193 +242,317 @@ InitStatus R3BCalifaCrystalCal2Cluster::ReInit()
-            // geometry. The reference angle will then have to be defined
-            // in relation to this reference distance: for example, 10° at
-            // 35 cm corresponds to ~6cm, setting a fDeltaAngleClust=10
-            // means that the gamma rays will be allowed to travel 6 cm in
-            // the CsI, no matter the position of the crystal they hit.
-            if (((vref.Angle(vhit)) * ((vref.Mag() + vhit.Mag()) / (35. * 2.))) < fDeltaAngleClust)
+
+            if (thisCryId <= fTotalCrystals && !isInside(usedCrystals, thisCryId) && thisEnergy < fSaturation &&
+                !fSimulation)
             {
-                takeCrystalInCluster = true;
+
+                angles = fCalifaGeo->GetAngles(thisCryId) - fCalifatoTargetPos;
+
+                if (InsideClusterWindow(mother_angles, angles))
+                    addCrystal2Cluster(&cluster, allCrystalVec.at(j), "gamma", &usedCrystals, fTotalCrystals);
             }
-            break;
-        case ROUND_SCALED: // round window scaled with energy
-            // The same as before but the angular window is scaled
-            // according to the energy of the hit in the higher energy
-            // crystal. It needs a parameter that should be calibrated.
+
+            /* ------- Simulation ------- */
+            if (!isInside(usedCrystals, thisCryId) && fSimulation)
             {
-                Double_t fDeltaAngleClustScaled = fDeltaAngleClust * (ref->GetEnergy() * energyFactor);
-                if (((vref.Angle(vhit)) * ((vref.Mag() + vhit.Mag()) / (35. * 2.))) < fDeltaAngleClustScaled)
-                {
-                    takeCrystalInCluster = true;
-                }
+
+                angles = fCalifaGeo->GetAngles(thisCryId) - fCalifatoTargetPos;
+
+                if (InsideClusterWindow(mother_angles, angles))
+                    addCrystal2Cluster(&cluster, allCrystalVec.at(j), "gamma", &usedCrystals, fTotalCrystals);
             }
-            break;
-        case CONE:
-            takeCrystalInCluster = vref.Angle(vhit) < fDeltaAngleClust;
-            break;
-        case PETAL:
-            takeCrystalInCluster = AngleToPetalId(vref) == AngleToPetalId(vhit);
-        case INVALID:
-        default:
-            throw std::runtime_error("R3BCalifaCrystalCal2Cluster: no clustering"
-                                     " algorithm selected.");
-            break;
+        }
+
+        AddHit(
+            cluster.crystalList, cluster.energy, cluster.nf, cluster.ns, cluster.theta, cluster.phi, cluster.time, 0);
+
+        RemoveUsedCrystals(
+            usedCrystals, allCrystalVec, protonCandidatesVec, gammaCandidatesVec, saturatedCandidatesVec);
     }
-    LOG(debug) << "returning R3BCalifaCrystalCal2Cluster::Match(" << ref->GetCrystalId() << ", " << hit->GetCrystalId()
-               << ")=" << takeCrystalInCluster << " with alg " << fClusterAlgorithmSelector;
 
-    return takeCrystalInCluster;
-}
+    /*------ Gamma Clusters ------- */
```

Comment:
This gamma-clustering algorithm is so-far the same as the one for proton. Only the variables are different. It would be easy to be understood if you can create a sub-routine for representing the both algorithm, that would be much nicer.

Reply:
There are subtle movements in data for both cases. I think this is clear as it is.